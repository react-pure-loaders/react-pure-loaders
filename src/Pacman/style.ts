import { keyframes } from '@emotion/react';
import { balls } from '../mixins';
import delay from '../utilities';
import { PRIMARY_COLOR } from '../variables';
import { Grid } from '../types';

const size = '25px';

const rotatePacmanHalfUp = keyframes({
  '0%': {
    transform: 'rotate(270deg)',
  },
  '100%': {
    transform: 'rotate(270deg)',
  },
  '50%': {
    transform: 'rotate(360deg)',
  },
});

const rotatePacmanHalfDown = keyframes({
  '0%': {
    transform: 'rotate(90deg)',
  },
  '100%': {
    transform: 'rotate(90deg)',
  },
  '50%': {
    transform: 'rotate(0deg)',
  },
});

const pacmanBalls = keyframes(
  {
    '100%': {
      transform: `translate(-${4 * parseInt(size, 10)}px, -${parseInt(size, 10) / 4}px)`,
    },
    '75%': {
      opacity: 0.7,
    },
  },
);

const pacmanDesign = (color?: string) => ({
  borderBottom: `${size} solid ${color || PRIMARY_COLOR}`,
  borderLeft: `${size} solid ${color || PRIMARY_COLOR}`,
  borderRadius: size,
  borderRight: `${size} solid transparent`,
  borderTop: `${size} solid ${color || PRIMARY_COLOR}`,
  height: '0px',
  width: '0px',
});

function ballPlacement() {
  const n = 3;
  let i = 0;
  const grid: Grid = {};
  for (; i < 4; i += 1) {
    grid[`> div:nth-of-type(${i + 2})`] = {
      animation: `${pacmanBalls} 1s ${delay('.33s', n, i)} infinite linear`,
    };
  }

  return grid;
}

export default (color?: string) => ({
  ...ballPlacement(),
  position: 'relative' as 'relative',
  ' > div:first-of-type': {
    ...pacmanDesign(color),
    animation: `${rotatePacmanHalfUp} 0.5s 0s infinite`,
    left: '-30px',
    position: 'relative' as 'relative',
  },

  ' > div:nth-of-type(2)': {
    ...pacmanDesign(color),
    animation: `${rotatePacmanHalfDown} 0.5s 0s infinite`,
    left: '-30px',
    marginTop: (-2 * parseInt(size, 10)),
    position: 'relative' as 'relative',
  },

  ' > div:nth-of-type(3), > div:nth-of-type(4), > div:nth-of-type(5), > div:nth-of-type(6)': {
    ...balls(color),
    height: '10px',
    left: '70px',
    position: 'absolute' as 'absolute',
    top: '25px',
    transform: `translate(0,-${parseInt(size, 10) / 4}px)`,
    width: '10px',
  },
});
