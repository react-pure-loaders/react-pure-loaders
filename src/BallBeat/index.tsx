/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallBeat({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div/>
                <div/>
                <div/>
            </div>
    );
  }

  return null;
}
