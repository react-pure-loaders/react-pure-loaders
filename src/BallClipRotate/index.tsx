/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallClipRotate({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)} data-testid="ball-clip-rotate-container">
                <div/>
            </div>
    );
  }

  return null;
}
