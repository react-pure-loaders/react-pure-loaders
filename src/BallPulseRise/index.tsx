/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallPulseRise({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div/>
                <div/>
                <div/>
                <div/>
                <div/>
            </div>
    );
  }

  return null;
}
