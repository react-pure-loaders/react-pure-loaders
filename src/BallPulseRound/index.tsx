/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallPulseRound({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div/>
            </div>
    );
  }

  return null;
}
