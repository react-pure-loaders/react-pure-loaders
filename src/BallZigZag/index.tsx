/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallZigZag({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div/>
                <div/>
            </div>
    );
  }

  return null;
}
