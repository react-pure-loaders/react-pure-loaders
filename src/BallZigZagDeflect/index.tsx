/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

export default function BallZigZagDeflect({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div/>
                <div/>
            </div>
    );
  }

  return null;
}
