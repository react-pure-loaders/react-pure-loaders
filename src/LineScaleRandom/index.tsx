/** @jsx jsx */
import { jsx } from '@emotion/react';
import style from './style';
import { BaseProps } from '../types';

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export default function LineScaleRandom({ color, loading }: BaseProps) {
  if (loading) {
    return (
            <div css={style(color)}>
                <div />
                <div />
                <div />
                <div />
            </div>
    );
  }

  return null;
}
