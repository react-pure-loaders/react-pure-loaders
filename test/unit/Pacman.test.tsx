import React from 'react';
import Chance from 'chance';
import { matchers } from '@emotion/jest';
import { render, cleanup } from '@testing-library/react';


import Pacman from '../../src/Pacman';
import { PRIMARY_COLOR } from '../../src/variables';

expect.extend(matchers);

const chance = new Chance();

describe('<Pacman>', () => {
    afterEach(cleanup);

    test('Pacman should match snapshot', () => {
        const { container } = render(<Pacman loading={true}/>);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('Pacman should have default color', () => {
        const { container } = render(<Pacman loading={true}/>);

        expect(container.firstChild).toHaveStyleRule('background-color', PRIMARY_COLOR, { target: '> div' });
    });

    test('Pacman should have given color', () => {
        const color = chance.color({ format: 'hex' });
        const { container } = render(<Pacman color={color} loading={true}/>);

        expect(container.firstChild).toHaveStyleRule('background-color', color, { target: '> div' });
    });

    test('Pacman should have no children', () => {
        const { container } = render(<Pacman loading={false}/>);

        expect(container.firstChild).toBeNull();
    });

    test('Pacman should have three children', () => {
        const { container } = render(<Pacman loading={true}/>);

        expect(container.querySelectorAll('div')).toHaveLength(6);
    });
});
