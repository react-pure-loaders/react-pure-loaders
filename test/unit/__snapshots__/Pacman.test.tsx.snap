// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pacman> Pacman should match snapshot 1`] = `
@keyframes animation-0 {
  100% {
    -webkit-transform: translate(-100px, -6.25px);
    -moz-transform: translate(-100px, -6.25px);
    -ms-transform: translate(-100px, -6.25px);
    transform: translate(-100px, -6.25px);
  }

  75% {
    opacity: 0.7;
  }
}

@keyframes animation-0 {
  100% {
    -webkit-transform: translate(-100px, -6.25px);
    -moz-transform: translate(-100px, -6.25px);
    -ms-transform: translate(-100px, -6.25px);
    transform: translate(-100px, -6.25px);
  }

  75% {
    opacity: 0.7;
  }
}

@keyframes animation-0 {
  100% {
    -webkit-transform: translate(-100px, -6.25px);
    -moz-transform: translate(-100px, -6.25px);
    -ms-transform: translate(-100px, -6.25px);
    transform: translate(-100px, -6.25px);
  }

  75% {
    opacity: 0.7;
  }
}

@keyframes animation-0 {
  100% {
    -webkit-transform: translate(-100px, -6.25px);
    -moz-transform: translate(-100px, -6.25px);
    -ms-transform: translate(-100px, -6.25px);
    transform: translate(-100px, -6.25px);
  }

  75% {
    opacity: 0.7;
  }
}

@keyframes animation-1 {
  0% {
    -webkit-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
    transform: rotate(270deg);
  }

  100% {
    -webkit-transform: rotate(270deg);
    -moz-transform: rotate(270deg);
    -ms-transform: rotate(270deg);
    transform: rotate(270deg);
  }

  50% {
    -webkit-transform: rotate(360deg);
    -moz-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes animation-2 {
  0% {
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
  }

  100% {
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
  }

  50% {
    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
}

.emotion-0 {
  position: relative;
}

.emotion-0>div:nth-of-type(2) {
  -webkit-animation: animation-0 1s -0.99s infinite linear;
  animation: animation-0 1s -0.99s infinite linear;
}

.emotion-0>div:nth-of-type(3) {
  -webkit-animation: animation-0 1s -0.66s infinite linear;
  animation: animation-0 1s -0.66s infinite linear;
}

.emotion-0>div:nth-of-type(4) {
  -webkit-animation: animation-0 1s -0.33s infinite linear;
  animation: animation-0 1s -0.33s infinite linear;
}

.emotion-0>div:nth-of-type(5) {
  -webkit-animation: animation-0 1s 0.00s infinite linear;
  animation: animation-0 1s 0.00s infinite linear;
}

.emotion-0>div:first-of-type {
  border-bottom: 25px solid #ffffff;
  border-left: 25px solid #ffffff;
  border-radius: 25px;
  border-right: 25px solid transparent;
  border-top: 25px solid #ffffff;
  height: 0px;
  width: 0px;
  -webkit-animation: animation-1 0.5s 0s infinite;
  animation: animation-1 0.5s 0s infinite;
  left: -30px;
  position: relative;
}

.emotion-0>div:nth-of-type(2) {
  border-bottom: 25px solid #ffffff;
  border-left: 25px solid #ffffff;
  border-radius: 25px;
  border-right: 25px solid transparent;
  border-top: 25px solid #ffffff;
  height: 0px;
  width: 0px;
  -webkit-animation: animation-2 0.5s 0s infinite;
  animation: animation-2 0.5s 0s infinite;
  left: -30px;
  margin-top: -50px;
  position: relative;
}

.emotion-0>div:nth-of-type(3),
.emotion-0>div:nth-of-type(4),
.emotion-0>div:nth-of-type(5),
.emotion-0>div:nth-of-type(6) {
  border-radius: 100%;
  height: 10px;
  margin: 2px;
  width: 10px;
  background-color: #ffffff;
  left: 70px;
  position: absolute;
  top: 25px;
  -webkit-transform: translate(0,-6.25px);
  -moz-transform: translate(0,-6.25px);
  -ms-transform: translate(0,-6.25px);
  transform: translate(0,-6.25px);
}

<div
  class="emotion-0"
>
  <div />
  <div />
  <div />
  <div />
  <div />
</div>
`;
